File: research/experiments/sweep/runs/opt-medium-kotlin-sonnet45/src/main/kotlin/optimization/VecOps.kt
package optimization

import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Pure vector arithmetic for n-dimensional optimization.
 * All operations return new arrays and never mutate inputs.
 */

/** Dot product */
fun dot(a: DoubleArray, b: DoubleArray): Double {
    require(a.size == b.size) { "Vectors must have same length" }
    return a.indices.sumOf { a[it] * b[it] }
}

/** Euclidean (L2) norm */
fun norm(v: DoubleArray): Double {
    return sqrt(v.sumOf { it * it })
}

/** Infinity norm (max absolute value) */
fun normInf(v: DoubleArray): Double {
    return v.maxOfOrNull { abs(it) } ?: 0.0
}

/** Scalar multiplication */
fun scale(v: DoubleArray, s: Double): DoubleArray {
    return DoubleArray(v.size) { v[it] * s }
}

/** Element-wise addition */
fun add(a: DoubleArray, b: DoubleArray): DoubleArray {
    require(a.size == b.size) { "Vectors must have same length" }
    return DoubleArray(a.size) { a[it] + b[it] }
}

/** Element-wise subtraction */
fun sub(a: DoubleArray, b: DoubleArray): DoubleArray {
    require(a.size == b.size) { "Vectors must have same length" }
    return DoubleArray(a.size) { a[it] - b[it] }
}

/** Element-wise negation */
fun negate(v: DoubleArray): DoubleArray {
    return scale(v, -1.0)
}

/** Deep copy */
fun clone(v: DoubleArray): DoubleArray {
    return v.copyOf()
}

/** Vector of n zeros */
fun zeros(n: Int): DoubleArray {
    return DoubleArray(n)
}

/** a + s*b (fused operation) */
fun addScaled(a: DoubleArray, b: DoubleArray, s: Double): DoubleArray {
    require(a.size == b.size) { "Vectors must have same length" }
    return DoubleArray(a.size) { a[it] + s * b[it] }
}

File: research/experiments/sweep/runs/opt-medium-python-sonnet45/src/vec_ops.py
"""
vec-ops — Pure vector arithmetic for n-dimensional optimization.

All operations return new lists and never mutate inputs.
"""

import math
from typing import List

Vector = List[float]


def dot(a: Vector, b: Vector) -> float:
    """Dot product of two vectors."""
    return sum(ai * bi for ai, bi in zip(a, b))


def norm(v: Vector) -> float:
    """Euclidean (L2) norm."""
    return math.sqrt(sum(x * x for x in v))


def norm_inf(v: Vector) -> float:
    """Infinity norm (max absolute value)."""
    if not v:
        return 0.0
    return max(abs(x) for x in v)


def scale(v: Vector, s: float) -> Vector:
    """Scalar multiplication."""
    return [x * s for x in v]


def add(a: Vector, b: Vector) -> Vector:
    """Element-wise addition."""
    return [ai + bi for ai, bi in zip(a, b)]


def sub(a: Vector, b: Vector) -> Vector:
    """Element-wise subtraction."""
    return [ai - bi for ai, bi in zip(a, b)]


def negate(v: Vector) -> Vector:
    """Element-wise negation."""
    return scale(v, -1.0)


def clone(v: Vector) -> Vector:
    """Deep copy."""
    return list(v)


def zeros(n: int) -> Vector:
    """Vector of n zeros."""
    return [0.0] * n


def add_scaled(a: Vector, b: Vector, s: float) -> Vector:
    """Compute a + s*b (fused operation)."""
    return [ai + s * bi for ai, bi in zip(a, b)]
